contracts/Rewards.sol

```diff
@@ -0,0 +1,212 @@
+  function isEligibleForRewards(address operator) public view returns (bool) {
+    return operatorRewards[operator].ineligibleUntil == 0;
+  }
+
+  /// @notice Return the time the operator's reward eligibility can be restored.
+  function rewardsEligibilityRestorableAt(address operator)
+    public
+    view
+    returns (uint256)
+  {
+    uint32 until = operatorRewards[operator].ineligibleUntil;
+    require(until != 0, "Operator already eligible");
+    return (uint256(until) + ineligibleOffsetStart);
+  }
+
+  /// @notice Return whether the operator is able
+  /// to restore their eligibility for rewards right away.
+  function canRestoreRewardEligibility(address operator)
+    public
+    view
+    returns (bool)
+  {
+    // solhint-disable-next-line not-rely-on-time
+    return rewardsEligibilityRestorableAt(operator) <= block.timestamp;
+  }
+
+  /// @notice Internal function for updating the global state of rewards.
+  function addRewards(uint96 rewardAmount, uint32 currentPoolWeight) internal {
+    require(currentPoolWeight >= 0, "No recipients in pool");
+
+    uint96 rewardWeight = uint96(currentPoolWeight - totalIneligibleWeight);
+
+    uint96 totalAmount = rewardAmount + rewardRoundingDust;
+    uint96 perWeightReward = totalAmount / rewardWeight;
+    uint96 newRoundingDust = totalAmount % rewardWeight;
+
+    globalRewardAccumulator += perWeightReward;
+    rewardRoundingDust = newRoundingDust;
+  }
+
+  /// @notice Internal function for updating the operator's reward state.
+  function updateOperatorRewards(address operator, uint32 newWeight) internal {
+    uint96 acc = globalRewardAccumulator;
+    OperatorRewards memory o = operatorRewards[operator];
+    if (o.ineligibleUntil == 0) {
+      // If operator is not ineligible, update their earned rewards
+      uint96 accruedRewards = (acc - o.accumulated) * uint96(o.weight);
+      o.available += accruedRewards;
+    } else {
+      // If ineligible, update total ineligible weight
+      uint32 iWeight = totalIneligibleWeight;
+      iWeight -= o.weight;
+      iWeight += newWeight;
+      totalIneligibleWeight = iWeight;
+    }
+    // In any case, update their accumulator and weight
+    o.accumulated = acc;
+    o.weight = newWeight;
+    operatorRewards[operator] = o;
+  }
+
+  /// @notice Set the amount of withdrawable tokens to zero
+  /// and return the previous withdrawable amount.
+  /// @dev Does not update the withdrawable amount,
+  /// but should usually be accompanied by an update.
+  function withdrawOperatorRewards(address operator)
+    internal
+    returns (uint96 withdrawable)
+  {
+    OperatorRewards storage o = operatorRewards[operator];
+    withdrawable = o.available;
+    o.available = 0;
+  }
+
+  /// @notice Set the given operators as ineligible for rewards.
+  /// The operators can restore their eligibility at the given time.
+  function setIneligible(address[] memory operators, uint256 until) internal {
```

Comment:
How hard (if possible at all) would be to modify the code so that if someone was marked as ineligible for rewards, other sortition pool members do not receive their share?

Group members may together sign a transaction saying who failed heartbeat and is ineligible for rewards. Ideally, this should not give them any higher rewards (notifier of the heartbeat failure TX is receiving a fixed reward from a separate pool). I am thinking about something like:
- `receiveApproval` received an approval for 1000 T,
- 5 T would go to 0xABC,
- 0xABC is ineligible so `receiveApproval` transfers only 1000-5=995 T from the funder.

Reply:
It would be very simple; we would simply remove the tracking of total ineligible weight.

However, this would effectively lose the rewards that would have been earned by ineligible operators, by making them inaccessible as nobody can withdraw them. We could permit the retrieval of those rewards by keeping the tracking of ineligible weight and accumulating all ineligible operators' rewards in "account zero" where someone with a special role (who?) can withdraw them or burn them formally.

Reply:
> where someone with a special role (who?) can withdraw them

I think `onlyOwner` and we will handle it in the application by allowing DAO (reward funder) to do whatever they want with this share.

Reply:
Got it 👍